### ISSUE TITLE
XGBClassifier, XGB booster, output_margin, base_score, etc prediction disrepancy

### ISSUE THREAD

**Issue by Yuan-Chang-UT** [dmlc/xgboost#4682]
Hi, I cannot understand how exactly XGB predictions are calculated at 3 phases.

I'll talk about 3 types of predictions i'm concerned about: XGBClassifier, XGBClassifier.get_booster() as booster, and booster.predict(output_margin=True) as output_margin predictions.

To my understanding, the sigmoid of the output_margin prediction is the booster.prediction, and it is always true as it should.

To my understanding, output_margin predictions should be the raw sums of the tree traversals, this holds when base_score is the 0.5 default, but when I set base_score to 0.4 for example in my example, I can't see how to calculate the output_margin prediction by hand with the trees.

My code is as follows:

I train a trivial model w/ one split and one tree to demonstrate.

```
xgb_params_sklearn = {
    "max_depth": 1,
    "n_estimators":1,
    "base_score": 0.5,
}

model = xgb.XGBClassifier()
model.set_params(**xgb_params_sklearn)
model.fit(x_train, y_train)
print(model.get_booster().get_dump(dump_format="json"))

fids = ['f0', 'f1', 'f2']
vals1 = [-99.0, 0, 0]
vals2 = [99.0, 0, 0]

df1 = pd.DataFrame([vals1], columns=fids)
df2 = pd.DataFrame([vals2], columns=fids)
dmat1 = xgb.DMatrix(df1)
dmat2 = xgb.DMatrix(df2)

print("HEAD")
print(df1.head())

booster = model.get_booster()
print("booster prediction, output_margin")
print(booster.predict(dmat1, output_margin=True))
print("booster prediction, ")
print(booster.predict(dmat1))
print("classifier prediction")
print(model.predict(df1))

print("booster prediction, output_margin")
print(booster.predict(dmat2, output_margin=True))
print("booster prediction, ")
print(booster.predict(dmat2))
print("classifier prediction")
print(model.predict(df2))
```

results with base_score = 0.5
```
{ "nodeid": 0, "depth": 0, "split": "f0", "split_condition": 1.45, "yes": 1, "no": 2, "missing": 1, 
"children": 
[{ "nodeid": 1, "leaf": -0.157895 },
{ "nodeid": 2, "leaf": 0.157895 }]}'

> HEAD
>      f0  f1  f2
> 0 -99.0   0   0

> booster prediction, output_margin
> [-0.15789475]

> booster prediction, 
> [0.46060812]

> classifier prediction
> [0]

> booster prediction, output_margin
> [0.15789475]

> booster prediction, 
> [0.5393919]

> classifier prediction
> [1]

```

As you can see the output_margin prediction is equal to the tree traversal sum, the two results go down to the left and right children which have values of -/+ 0.157895

sigmoid(-0.15789475) == 0.46060812, the sigmoid/logit relationship between the output_margin and booster prediction holds

**QUESTION 1: But how does the classifier predict 0 and 1, what maps the booster prediction to the classifier prediction?** Does it have to do with the objective hyperparameter? Can someone give me an example of how to get the classifier result from the booster result?

#
#
#

Also, other things cause me confusing when I change the base_score from the default of 0.5

results with base_score = 0.4

```
{ "nodeid": 0, "depth": 0, "split": "f0", "split_condition": 1.45, "yes": 1, "no": 2, "missing": 1, 
"children": 
[{ "nodeid": 1, "leaf": -0.0221739 },    
{"nodeid": 2, "leaf": 0.0332609 }]}'

> HEAD
>      f0  f1  f2
> 0 -99.0   0   0

> booster prediction, output_margin
> [-0.427639]

> booster prediction
> [0.39469028]

> classifier prediction
> [0]

> booster prediction, output_margin
> [-0.3722042]

> booster prediction
> [0.40800852]

> classifier prediction
> [0]
```

Changing the base_score changes the tree leaf values for a low amount of n_estimators to my understanding, after n_estimators is large enough the trees will converge despite base_score, to my understanding.

sigmoid(-0.427639) == 0.39469028, the sigmoid/logit relation still holds.

**QUESTION 2: But where does the output_margin outputs of -0.427639 and -0.3722042 come from, I can see they are centered at -0.4 and the base_score is 0.4, but how exactly would one calculate this by hand from the trees?** They don't seem to be the raw tree traversal sums anymore, unlike w/ base_score =0.5

 **QUESTION 3: Is it entirely possible to calculate this with just the information encoded in the trees?**

**Comment by Yuan-Chang-UT** [comment]
@hcho3 can you help me out again? Or tag someone who can?

**Comment by Yuan-Chang-UT** [comment]
I figured out half of it at least, the booster prediction is equal to:


![image](https://user-images.githubusercontent.com/22758979/61814792-b403f480-adfd-11e9-9de6-f4b403cec43d.png)

With the default of base_score == 0.5, logit(0.5) == 0 as Sigmoid(0) == 0.5

**Comment by hcho3** [comment]
Indeed, the prediction is computed by `sigmoid(sum(leaf_outputs) + logit(base_score))`.
